paho/pinger.go

```diff
@@ -2,121 +2,45 @@ package paho
 
 import (
 	"fmt"
-	"net"
-	"sync"
-	"sync/atomic"
 	"time"
 
 	"github.com/eclipse/paho.golang/packets"
 )
 
-// PingFailHandler is a type for the function that is invoked
-// when we have sent a Pingreq to the server and not received
-// a Pingresp within 1.5x our pingtimeout
-type PingFailHandler func(error)
 
-// Pinger is an interface of the functions for a struct that is
-// used to manage sending PingRequests and responding to
-// PingResponses
-// Start() takes a net.Conn which is a connection over which an
-// MQTT session has already been established, and a time.Duration
-// of the keepalive setting passed to the server when the MQTT
-// session was established.
-// Stop() is used to stop the Pinger
-// PingResp() is the function that is called by the Client when
-// a PingResponse is received
-// SetDebug() is used to pass in a Logger to be used to log debug
-// information, for example sharing a logger with the main client
-type Pinger interface {
-	Start(net.Conn, time.Duration)
-	Stop()
-	PingResp()
-	SetDebug(Logger)
-}
-
-// PingHandler is the library provided default Pinger
-type PingHandler struct {
-	mu              sync.Mutex
-	lastPing        time.Time
-	conn            net.Conn
-	stop            chan struct{}
-	pingFailHandler PingFailHandler
-	pingOutstanding int32
-	debug           Logger
-}
+func StartPinger(
+	cPingresp chan *packets.Pingresp, 
+	cOutgoingPackets chan *packets.ControlPacket, 
+	cStop chan struct{},
+	cError chan error,
+	pt time.Duration,
+	debug Logger,
+) {
+	debug.Println("Starting pinger")
+	sendNextPing := time.After(pt)
+	pingExpired := make(<-chan time.Time)
 
-// DefaultPingerWithCustomFailHandler returns an instance of the
-// default Pinger but with a custom PingFailHandler that is called
-// when the client has not received a response to a PingRequest
-// within the appropriate amount of time
-func DefaultPingerWithCustomFailHandler(pfh PingFailHandler) *PingHandler {
-	return &PingHandler{
-		pingFailHandler: pfh,
-		debug:           NOOPLogger{},
-	}
-}
-
-// Start is the library provided Pinger's implementation of
-// the required interface function()
-func (p *PingHandler) Start(c net.Conn, pt time.Duration) {
-	p.mu.Lock()
-	p.conn = c
-	p.stop = make(chan struct{})
-	p.mu.Unlock()
-	checkTicker := time.NewTicker(pt / 4)
-	defer checkTicker.Stop()
 	for {
 		select {
-		case <-p.stop:
+		case <- cStop:
+			debug.Println("Pinger stopped")
+			return
+		case <- sendNextPing:
+		    debug.Println("Sending next ping")
+			go func() {
+				cOutgoingPackets <- packets.NewControlPacket(packets.PINGREQ)
+			}()
+			sendNextPing = nil
+			pingExpired = time.After(pt / 2)
```

Comment:
I'm not sure on assigning channels in the select statement like this. It looks like it will work without an problem, but it might be a smell.

Reply:
This was one of the big ideas I got from that advanced concurrency talk It saves so much complexity

Reply:
Do you have a timestamp (if it's not too much trouble)